confidant/services/certificatemanager.py

```diff
@@ -0,0 +1,397 @@
+        # this look weirdly javascript-like.
+        cert = x509.CertificateBuilder(
+        ).subject_name(
+            subject,
+        ).issuer_name(
+            issuer,
+        ).public_key(
+            key.public_key(),
+        ).serial_number(
+            x509.random_serial_number(),
+        ).not_valid_before(
+            datetime.datetime.utcnow(),
+        ).not_valid_after(
+            datetime.datetime.utcnow() + datetime.timedelta(days=_validity),
+        )
+        if san:
+            dns_names = self.encode_san_dns_names(san)
+            cert = cert.add_extension(
+                x509.SubjectAlternativeName(dns_names),
+                critical=False,
+            )
+        return cert.sign(key, hashes.SHA256(), default_backend())
+
+    def encode_certificate(self, cert):
+        """
+        Return the PEM string encoded version of the certificate object.
+        """
+        return cert.public_bytes(serialization.Encoding.PEM)
+
+    def issue_certificate(self, csr, validity):
+        """
+        Given a PEM encoded csr, and a validity for the certificate (in number
+        of days), issue a certificate from ACM Private CA, and return the ARN
+        of the issued certificate.
+        """
+        client = confidant.clients.get_boto_client('acm-pca')
+        response = client.issue_certificate(
+            CertificateAuthorityArn=self.settings['arn'],
+            Csr=csr,
+            SigningAlgorithm=self.settings['signing_algorithm'],
+            Validity={
+                'Value': min(validity, self.settings['max_validity_days']),
+                'Type': 'DAYS',
+            },
+            # Quick/easy idempotent token is just a hash of the csr itself. The
+            # token must be 36 chars or less.
+            IdempotencyToken=hashlib.sha256(csr).hexdigest()[:36],
+        )
+        return response['CertificateArn']
+
+    def _get_cached_certificate_with_key(self, cache_id):
+        """
+        For the cache id, get the cached response, or, if another thread is in
+        the process of issuing the same certificate, wait for the other thread
+        to populate the cache.
+        """
+        if not self.settings['certificate_use_cache']:
+            return {}
+        item = self.cache.get(cache_id)
+        # We're the first thread attempting to get this certificate
+        if not item:
+            return {}
+        # A certificate hasn't been issued yet, but since the cache id exists,
+        # another thread has requested the certificate.
+        if not item.response:
+            # Wait for response
+            while True:
+                # keep waiting while the lock is held.
+                if item.lock:
+                    logging.debug(
+                        'Sleeping in _get_cached_certificate_with_key'
+                        ' for {}'.format(cache_id)
+                    )
+                    time.sleep(.100)
```

Comment:
ooc -- do you think it's worth adding a max loop iteration count as a fail safe to prevent an infinite loop if the cache fails to unlock?

Reply:
gunicorn has a max timeout per-thread, but maybe it's a good idea to break before that, as I guess it's possible that not all wsgi workers do, and maybe someone's running this in a wsgi worker that isn't gunicorn.